File: Coding Interview Workshops (CIW)/Week 03 - Stacks and Queues/Code/leetcode.py
from collections import deque
"""
APC

1. Assumptions
 - int in stack
 - 100 calls to push, pop, top and empty
 - use deque (doubly linked list)


2. Plan 

[9,0,7,8]
 - push (1) -> [1,9,0,7,8]
 - pop () -> [9,0,7,8]

 [] [9,0,7,8] -> 1 -> [1] [9,0,7,8] -> [1,9,0,7,8] []
                   -> [9,0,7,8,1] -> [0,7,8,1,9], [7,8,1,9,0], [8,1,9,0,7], [1,9,0,7,8]

 - enqueue (1) -> [9,0,7,8,1]
 - dequeue () -> [0,7,8,1]


3. Code


"""

class MyStack(object):
    def __init__(self):
        self.q1 = deque()
        self.q2 = deque()
        self.in_q1 = True

    def push(self, x):
        """
        :type x: int
        :rtype: None
        """
        if (self.in_q1):
            self.q2.append(x)
            while (len(self.q1)>0):
                y = self.q1.popleft()
                self.q2.append(y)
            self.in_q1 = False
        else:
            self.q1.append(x)
            while (len(self.q2)>0):
                y = self.q2.popleft()
                self.q1.append(y)
            self.in_q1 = True

        

    def pop(self):
        """
        :rtype: int
        """
        if (self.in_q1):
            return self.q1.popleft()
        else:
            return self.q2.popleft()
        
        

    def top(self):
        """
        :rtype: int
        """
        if (self.in_q1):
            return self.q1[0]
        else:
            return self.q2[0]
        

    def empty(self):
        """
        :rtype: bool
        """
        if (self.in_q1):
            if len(self.q1) == 0:
                return True
            else:
                return False
        else:
            if len(self.q2) == 0:
                return True
            else:
                return False
        


# Your MyStack object will be instantiated and called as such:
# obj = MyStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.empty()

File: Coding Interview Workshops (CIW)/Week 03 - Stacks and Queues/Code/leetcode.js
/**
    APC

    1. Assumptions
    - int in stack
    - 100 calls to push, pop, top and empty
    - use deque (doubly linked list)


    2. Plan 

    [9,0,7,8]
    - push (1) -> [1,9,0,7,8]
    - pop () -> [9,0,7,8]

    [] [9,0,7,8] -> 1 -> [1] [9,0,7,8] -> [1,9,0,7,8] []
                    -> [9,0,7,8,1] -> [0,7,8,1,9], [7,8,1,9,0], [8,1,9,0,7], [1,9,0,7,8]

    - enqueue (1) -> [9,0,7,8,1]
    - dequeue () -> [0,7,8,1]


    3. Code
    REMEMBER: enqueue is push, dequeue is shift because we're using a list to   implement the queue

 */

    var MyStack = function() {
        this.q1 = [];
        this.q2 = [];
        this.inQ1 = true;
    };
    
    /** 
     * @param {number} x
     * @return {void}
     */
    MyStack.prototype.push = function(x) {
        if (this.inQ1) {
            this.q2.push(x);
            while (this.q1.length > 0) {
                this.q2.push(this.q1.shift());
            }
            this.inQ1 = false;
        } else {
            this.q1.push(x);
            while (this.q2.length > 0) {
                this.q1.push(this.q2.shift());
            }
            this.inQ1 = true;
        }
    };
    
    /**
     * @return {number}
     */
    MyStack.prototype.pop = function() {
        if (this.inQ1) 
            return this.q1.shift();
        else 
            return this.q2.shift();
        
    };
    
    /**
     * @return {number}
     */
    MyStack.prototype.top = function() {
        if (this.inQ1) 
            return this.q1[0];
        else 
            return this.q2[0];
        
    };
    
    /**
     * @return {boolean}
     */
    MyStack.prototype.empty = function() {
        if (this.inQ1) 
            if (this.q1.length == 0)
                return true;
            else
                return false;
        else 
            if (this.q2.length == 0)
                return true;
            else
                return false;
    };
    
    /** 
     * Your MyStack object will be instantiated and called as such:
     * var obj = new MyStack()
     * obj.push(x)
     * var param_2 = obj.pop()
     * var param_3 = obj.top()
     * var param_4 = obj.empty()
     */